ibis/expr/types/relations.py

```diff
@@ -990,9 +993,35 @@ def group_by(
         """
         from ibis.expr.types.groupby import GroupedTable
 
+        def partition_groups(*by, **key_exprs):
+            grouping_sets = []
+            rollups = []
+            cubes = []
+            groups = []
+            for b in by:
+                if isinstance(b, Rollup):
+                    rollups.append(b.expand(self))
+                elif isinstance(b, Cube):
+                    cubes.append(b.expand(self))
+                elif isinstance(b, GroupingSets):
+                    grouping_sets.append(b.expand(self))
+                else:
+                    groups.extend(self.bind(b))
+
+            groups.extend(self.bind(**key_exprs))
+            return groups, grouping_sets, rollups, cubes
+
         by = tuple(v for v in by if v is not None)
-        groups = self.bind(*by, **key_exprs)
-        return GroupedTable(self, groups)
+        groups, grouping_sets, rollups, cubes = partition_groups(*by, **key_exprs)
+        if not (groups or grouping_sets or rollups or cubes):
+            raise com.IbisInputError("No grouping keys provided")
```

Comment:
Previously we were using an `Annotated` type hint to enforce that `groups` was non-empty, but we can't do that anymore because it **can** be empty if the user provides at least one grouping set, rollup, or cube and we don't have a way to spell that kind of relationship in the type system (nor do I think we need to build that right now).